man/res_user.5
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RES_USER 1"
.TH RES_USER 1 "2014-09-13" "Clockwork v2.3.0" "Clockwork Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
res_user \- Clockwork Resource Type for system users
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The user resource manages system user accounts.
.SH "ATTRIBUTES"
.IX Header "ATTRIBUTES"
.IP "\fBuid\fR" 4
.IX Item "uid"
Numeric user \s-1ID\s0 of this account.
.IP "\fBusername\fR" 4
.IX Item "username"
The username for this account.
.IP "\fBgid\fR" 4
.IX Item "gid"
Numeric group \s-1ID\s0 for this acount's primary group.
.IP "\fBhome\fR" 4
.IX Item "home"
Path to this user's home directory.  This attribute only sets
the home directory path in /etc/password.  To create home directories
for new users, see the \fBmakehome\fR / \fBskeleton\fR attribute.
.IP "\fBpresent\fR" 4
.IX Item "present"
Whether or not this account should exist on the system.
.IP "\fBlocked\fR" 4
.IX Item "locked"
Whether or not this account should be locked and therefore unable
to log in.  Locking an account does not destroy password information;
a locked account can be unlocked later and the original password will
work.
.IP "\fBcomment\fR" 4
.IX Item "comment"
.PD 0
.IP "\fBgecos\fR" 4
.IX Item "gecos"
.PD
A comment describing the purpose of this account, commonly refered to
as the \s-1GECOS\s0 field.  See \fBpasswd\fR(5) for more details.
.Sp
Clockwork does not attempt to interpret the \s-1GECOS\s0 field like the
\&\fBchfn\fR(1) utility does.
.IP "\fBshell\fR" 4
.IX Item "shell"
The path to the login shell for this account.  No validation is done
on this shell, since specifying non-existent or unapproved login shells
is one way of locking accounts out of a system.
.IP "\fBpassword\fR" 4
.IX Item "password"
.PD 0
.IP "\fBpwhash\fR" 4
.IX Item "pwhash"
.PD
The (encrypted) password for this account.  Details of encryption depend
on the specific platform and system configuration.  Clockwork does not
attempt to encrypt or decrypt passwords for users.
.IP "\fBchangepw\fR" 4
.IX Item "changepw"
Whether or not Clockwork should change the passowrd of an existing user
account, according to the value of \fBpassword\fR.  The primary use of
this attribute is for provisioning new accounts with default initial
passwords, but not forcibly reseting the password during subsequent
runs.
.IP "\fBpwmin\fR" 4
.IX Item "pwmin"
Minimum number of days between password changes.
.IP "\fBpwmax\fR" 4
.IX Item "pwmax"
Maximum age of the account password (in days).
.IP "\fBpwwarn\fR" 4
.IX Item "pwwarn"
Number of days before the password expires (according to \fBpwmax\fR)
to start warning the user that they should change their password.
.IP "\fBinact\fR" 4
.IX Item "inact"
Number of days after the password expires (according to \fBpwmax\fR)
before the account is reversibly disabled.
.IP "\fBexpiry\fR" 4
.IX Item "expiry"
.PD 0
.IP "\fBexpiration\fR" 4
.IX Item "expiration"
.PD
When the account expires.  This value is specified as the number of days
since Jan 1, 1970, and is not related to the \fBpwmax\fR, \fBpwmin\fR or
\&\fBinact\fR attributes.
.IP "\fBskeleton\fR" 4
.IX Item "skeleton"
.PD 0
.IP "\fBmakehome\fR" 4
.IX Item "makehome"
.PD
This attribute enables home directory creation for new users.  Valid values
include \*(L"yes\*(R", \*(L"no\*(R" or a path to a skeleton directory.  If \*(L"yes\*(R" is used,
then the system default of /etc/skel is used as a template to create the
new home directory.  Otherwise, the path given is used.
.Sp
Home directory creation is only effective if the account does not already
exist on the local system; Clockwork will not overwrite or otherwise meddle
with a pre-existing home directory.
.Sp
See the \fB\s-1EXAMPLES\s0\fR section for more information.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Removing and Locking Accounts"
.IX Subsection "Removing and Locking Accounts"
To remove a user account:
.PP
.Vb 3
\&    user "jim" {
\&        present: "no"
\&    }
.Ve
.PP
To keep the account around (including its original password),
but deny any and all login access to it:
.PP
.Vb 3
\&    user "jim" {
\&        locked: "yes"
\&    }
.Ve
.SS "Password Expiration"
.IX Subsection "Password Expiration"
The following example implements a password expiration policy:
.PP
.Vb 4
\&    user "eva" {
\&      pwmin:  1   # After password change, user must
\&                  # wait at least 1 day to change their
\&                  # password again.
\&
\&      pwmax:  45  # Passwords *must* be changed after
\&                  # 45 days.
\&
\&      pwwarn: 10  # Warn the user 10 days before their
\&                  # password expires that they will have
\&                  # to change it soon.  Under normal
\&                  # circumstances, this will be 35 days
\&                  # after they change it.
\&
\&      inact:  5   # Give the users 5 days after their
\&                  # password expires (50 days after they
\&                  # change it) before disabling the
\&                  # account.
\&    }
.Ve
.PP
If you specify password aging parameters, make sure you don't also set
the account password via the \fBpassword\fR / \fBpwhash\fR and
\&\fBchangepw\fR attributes.
.SS "Creating Home Directories"
.IX Subsection "Creating Home Directories"
The user resource can create the home directory for newly created
users if the \fBmakehome\fR or \fBskeleton\fR attributes are set.
.PP
Create the user \*(L"bob\*(R", and model his home directory after /etc/skel:
.PP
.Vb 6
\&    user "bob" {
\&        uid:      1006
\&        gid:      1001   # users, defined elsewhere
\&        home:     "/home/guests/bob"
\&        makehome: "yes"
\&    }
.Ve
.PP
If you want to use a different skeleton directory (i.e. for system
administrator accounts), specify its path in the \fBmakehome\fR
attribute:
.PP
.Vb 6
\&    user "james" {
\&      uid:       7001,
\&      gid:       1002,   # admins, defined elsewhere
\&      home:      "/home/james"
\&      makehome:  "/etc/skel.admin"
\&    }
.Ve
.PP
\&\fBNote:\fR
The \fBmakehome\fR and \fBskeleton\fR attributes are synonymous.
The last example is probably more clearly expressed as:
.PP
.Vb 6
\&    user "james" {
\&      uid:       7001,
\&      gid:       1002,   # admins, defined elsewhere
\&      home:      "/home/james"
\&      skeleton:  "/etc/skel.admin"
\&    }
.Ve
.SH "CAVEATS"
.IX Header "CAVEATS"
.SS "1. Removing Home Directories"
.IX Subsection "1. Removing Home Directories"
Clockwork does not remove a user's home directory when it removes
the account.  This is by design, to ensure that important data
is not lost when users depart.  If you want to remove the home
directory, use \fBres_file\fR(5) in concert with \fBres_user\fR.
.SS "2. Dependencies"
.IX Subsection "2. Dependencies"
For flexibility's sake, Clockwork does not create implicit dependencies
on users for  their primary group.  If you specify that a user's primary \s-1GID\s0
should be 177, Clockwork will happily comply, even if there is no group
defined on the system (either through policy or externally) with that \s-1GID.\s0
.PP
This should not break normal operation of the system, but it will lead to
some strangeness when dealing with files owned by that user.
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
See Caveat #2.
.SH "AUTHOR"
.IX Header "AUTHOR"
Clockwork was designed and written by James Hunt.


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




res_user.c
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fts.h>
#include <fcntl.h>
#include <unistd.h>

#include "res_user.h"
#include "pack.h"


#define RES_USER_PACK_PREFIX "res_user::"
#define RES_USER_PACK_OFFSET 10

/* Pack Format for res_user structure:
     L - ru_enf
     a - ru_name
     a - ru_passwd
     L - ru_uid
     L - ru_gid
     a - ru_gecos
     a - ru_shell
     a - ru_dir
     C - ru_mkhome
     a - ru_skel
     C - ru_locked
     L - ru_pwmin
     L - ru_pwmax
     L - ru_pwwarn
     L - ru_inact
     L - ru_expire
 */
#define RES_USER_PACK_FORMAT "LaaLLaaaCaCLLLLL"


static int _res_user_diff(struct res_user *ru);
static unsigned char _res_user_home_exists(struct res_user *ru);
static int _res_user_populate_home(const char *home, const char *skel, uid_t uid, gid_t gid);

/*****************************************************************/

static int _res_user_diff(struct res_user *ru)
{
	assert(ru);

	unsigned char locked = (ru->ru_sp->sp_pwdp && *(ru->ru_sp->sp_pwdp) == '!') ? 1 : 0;

	ru->ru_diff = RES_USER_NONE;

	if (res_user_enforced(ru, NAME) && strcmp(ru->ru_name, ru->ru_pw->pw_name) != 0) {
		ru->ru_diff |= RES_USER_NAME;
	}

	if (res_user_enforced(ru, PASSWD) && strcmp(ru->ru_passwd, ru->ru_sp->sp_pwdp) != 0) {
		ru->ru_diff |= RES_USER_PASSWD;
	}

	if (res_user_enforced(ru, UID) && ru->ru_uid != ru->ru_pw->pw_uid) {
		ru->ru_diff |= RES_USER_UID;
	}

	if (res_user_enforced(ru, GID) && ru->ru_gid != ru->ru_pw->pw_gid) {
		ru->ru_diff |= RES_USER_GID;
	}

	if (res_user_enforced(ru, GECOS) && strcmp(ru->ru_gecos, ru->ru_pw->pw_gecos) != 0) {
		ru->ru_diff |= RES_USER_GECOS;
	}

	if (res_user_enforced(ru, DIR) && strcmp(ru->ru_dir, ru->ru_pw->pw_dir) != 0) {
		ru->ru_diff |= RES_USER_DIR;
	}

	if (res_user_enforced(ru, SHELL) && strcmp(ru->ru_shell, ru->ru_pw->pw_shell) != 0) {
		ru->ru_diff |= RES_USER_SHELL;
	}

	if (ru->ru_mkhome == 1 && res_user_enforced(ru, MKHOME) && _res_user_home_exists(ru) == 0) {
		ru->ru_diff |= RES_USER_MKHOME;
	}

	if (res_user_enforced(ru, PWMIN) && ru->ru_pwmin != ru->ru_sp->sp_min) {
		ru->ru_diff |= RES_USER_PWMIN;
	}

	if (res_user_enforced(ru, PWMAX) && ru->ru_pwmax != ru->ru_sp->sp_max) {
		ru->ru_diff |= RES_USER_PWMAX;
	}

	if (res_user_enforced(ru, PWWARN) && ru->ru_pwwarn != ru->ru_sp->sp_warn) {
		ru->ru_diff |= RES_USER_PWWARN;
	}

	if (res_user_enforced(ru, INACT) && ru->ru_inact != ru->ru_sp->sp_inact) {
		ru->ru_diff |= RES_USER_INACT;
	}

	if (res_user_enforced(ru, EXPIRE) && ru->ru_expire != ru->ru_sp->sp_expire) {
		ru->ru_diff |= RES_USER_EXPIRE;
	}

	if (res_user_enforced(ru, LOCK) && ru->ru_lock != locked) {
		ru->ru_diff |= RES_USER_LOCK;
	}

	return 0;
}

static unsigned char _res_user_home_exists(struct res_user *ru)
{
	assert(ru);
	assert(ru->ru_dir);

	struct stat st;
	if (stat(ru->ru_dir, &st) == -1) {
		return 0; /* 0 = false */
	}

	return (S_ISDIR(st.st_mode) ? 1 : 0); /* 1 = true; 0 = false */
}

static int _res_user_populate_home(const char *home, const char *skel, uid_t uid, gid_t gid)
{
	FTS *fts;
	FTSENT *ent;
	char *path_argv[2] = { strdup(skel), NULL };
	char *skel_path, *home_path;
	int skel_fd, home_fd;
	mode_t mode;
	char buf[8192];
	size_t nread;

	fts = fts_open(path_argv, FTS_PHYSICAL, NULL);
	if (!fts) {
		free(path_argv[0]);
		return -1;
	}

	while ( (ent = fts_read(fts)) != NULL ) {
		if (strcmp(ent->fts_accpath, ent->fts_path) != 0) {
			home_path = string("%s/%s", home, ent->fts_accpath);
			skel_path = string("%s/%s", skel, ent->fts_accpath);
			mode = ent->fts_statp->st_mode & 0777;

			if (S_ISREG(ent->fts_statp->st_mode)) {
				skel_fd = open(skel_path, O_RDONLY);
				home_fd = creat(home_path, mode);

				if (skel_fd >= 0 && home_fd >= 0) {
					chown(home_path, uid, gid);
					while ((nread = read(skel_fd, buf, 8192)) > 0)
						write(home_fd, buf, nread);
				}

			} else if (S_ISDIR(ent->fts_statp->st_mode)) {
				if (mkdir(home_path, mode) == 0) {
					chown(home_path, uid, gid);
					chmod(home_path, 0755);
				}
			}

			free(home_path);
			free(skel_path);
		}
	}

	free(path_argv[0]);
	return 0;
}

/*****************************************************************/

void* res_user_new(const char *key)
{
	struct res_user *ru;

	ru = xmalloc(sizeof(struct res_user));

	ru->ru_uid    = -1;
	ru->ru_uid    = -1;
	ru->ru_mkhome = 0;
	ru->ru_lock   = 1;
	ru->ru_pwmin  = 0;
	ru->ru_pwmax  = 0;
	ru->ru_pwwarn = 0;
	ru->ru_inact  = 0;
	ru->ru_expire = 0;

	ru->ru_enf = RES_USER_NONE;
	ru->ru_diff = RES_USER_NONE;

	list_init(&ru->res);

	ru->ru_name   = NULL;
	ru->ru_passwd = NULL;
	ru->ru_gecos  = NULL;
	ru->ru_dir    = NULL;
	ru->ru_shell  = NULL;
	ru->ru_skel   = NULL;

	ru->ru_pw     = NULL;
	ru->ru_sp     = NULL;

	if (key) {
		res_user_set(ru, "username", key);
		ru->key = string("res_user:%s", key);
	} else {
		ru->key = NULL;
	}

	return ru;
}

void res_user_free(void *res)
{
	struct res_user *ru = (struct res_user*)(res);

	if (ru) {
		list_del(&ru->res);

		free(ru->ru_name);
		free(ru->ru_passwd);
		free(ru->ru_gecos);
		free(ru->ru_dir);
		free(ru->ru_shell);
		free(ru->ru_skel);

		free(ru->key);
	}
	free(ru);
}

const char* res_user_key(const void *res)
{
	const struct res_user *ru = (struct res_user*)(res);
	assert(ru);

	return ru->key;
}

int res_user_norm(void *res) { return 0; }

int res_user_set(void *res, const char *name, const char *value)
{
	struct res_user *ru = (struct res_user*)(res);
	assert(ru);

	if (strcmp(name, "uid") == 0) {
		ru->ru_uid = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_UID;

	} else if (strcmp(name, "gid") == 0) {
		ru->ru_gid = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_GID;

	} else if (strcmp(name, "username") == 0) {
		free(ru->ru_name);
		ru->ru_name = strdup(value);
		ru->ru_enf |= RES_USER_NAME;

	} else if (strcmp(name, "home") == 0) {
		free(ru->ru_dir);
		ru->ru_dir = strdup(value);
		ru->ru_enf |= RES_USER_DIR;

	} else if (strcmp(name, "present") == 0) {
		if (strcmp(value, "no") != 0) {
			ru->ru_enf ^= RES_USER_ABSENT;
		} else {
			ru->ru_enf |= RES_USER_ABSENT;
		}

	} else if (strcmp(name, "locked") == 0) {
		ru->ru_lock = strcmp(value, "no") ? 1 : 0;
		ru->ru_enf |= RES_USER_LOCK;

	} else if (strcmp(name, "gecos") == 0 || strcmp(name, "comment") == 0) {
		free(ru->ru_gecos);
		ru->ru_gecos = strdup(value);
		ru->ru_enf |= RES_USER_GECOS;

	} else if (strcmp(name, "shell") == 0) {
		free(ru->ru_shell);
		ru->ru_shell = strdup(value);
		ru->ru_enf |= RES_USER_SHELL;

	} else if (strcmp(name, "pwhash") == 0 || strcmp(name, "password") == 0) {
		free(ru->ru_passwd);
		ru->ru_passwd = strdup(value);
		ru->ru_enf |= RES_USER_PASSWD;

	} else if (strcmp(name, "pwmin") == 0) { // FIXME: need better key
		ru->ru_pwmin = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_PWMIN;

	} else if (strcmp(name, "pwmax") == 0) { // FIXME: need better key
		ru->ru_pwmax = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_PWMAX;

	} else if (strcmp(name, "pwwarn") == 0) { // FIXME: need better key
		ru->ru_pwwarn = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_PWWARN;

	} else if (strcmp(name, "inact") == 0) { // FIXME: need better key
		ru->ru_inact = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_INACT;

	} else if (strcmp(name, "expiry") == 0 || strcmp(name, "expiration") == 0) {
		ru->ru_expire = strtoll(value, NULL, 10);
		ru->ru_enf |= RES_USER_EXPIRE;

	} else if (strcmp(name, "skeleton") == 0 || strcmp(name, "makehome") == 0) {
		ru->ru_enf |= RES_USER_MKHOME;
		xfree(ru->ru_skel);
		ru->ru_mkhome = (strcmp(value, "no") ? 1 : 0);

		if (!ru->ru_mkhome) { return 0; }
		ru->ru_skel = strdup(strcmp(value, "yes") == 0 ? "/etc/skel" : value);

	}

	return 0;
}

int res_user_stat(void *res, const struct resource_env *env)
{
	struct res_user *ru = (struct res_user*)(res);
	assert(ru);
	assert(env);
	assert(env->user_pwdb);
	assert(env->user_spdb);

	ru->ru_pw = pwdb_get_by_name(env->user_pwdb, ru->ru_name);
	ru->ru_sp = spdb_get_by_name(env->user_spdb, ru->ru_name);
	if (!ru->ru_pw || !ru->ru_sp) { /* new account */
		ru->ru_diff = ru->ru_enf;
		return 0;
	}

	return _res_user_diff(ru);
}

struct report* res_user_fixup(void *res, int dryrun, const struct resource_env *env)
{
	struct res_user *ru = (struct res_user*)(res);
	assert(ru);
	assert(env);
	assert(env->user_pwdb);
	assert(env->user_spdb);

	struct report *report;
	char *action;

	int new_user = 0;

	report = report_new("User", ru->ru_name);

	/* Remove the user if RES_USER_ABSENT */
	if (res_user_enforced(ru, ABSENT)) {
		if (ru->ru_pw || ru->ru_sp) {
			action = string("remove user");

			if (dryrun) {
				report_action(report, action, ACTION_SKIPPED);
				return report;
			}

			if ((ru->ru_pw && pwdb_rm(env->user_pwdb, ru->ru_pw) != 0)
			 || (ru->ru_sp && spdb_rm(env->user_spdb, ru->ru_sp) != 0)) {
				report_action(report, action, ACTION_FAILED);
			} else {
				report_action(report, action, ACTION_SUCCEEDED);
			}
		}

		return report;
	}

	if (!ru->ru_pw || !ru->ru_sp) {
		action = string("create user");
		new_user = 1;

		if (!ru->ru_pw) { ru->ru_pw = pwdb_new_entry(env->user_pwdb, ru->ru_name, ru->ru_uid, ru->ru_gid); }
		if (!ru->ru_sp) { ru->ru_sp = spdb_new_entry(env->user_spdb, ru->ru_name); }

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (ru->ru_pw && ru->ru_sp) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
			return report;
		}
	}

	if (res_user_different(ru, PASSWD)) {
		if (new_user) {
			action = string("set user password");
		} else {
			action = string("change user password");
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(ru->ru_pw->pw_passwd);
			ru->ru_pw->pw_passwd = strdup("x");

			xfree(ru->ru_sp->sp_pwdp);
			ru->ru_sp->sp_pwdp = strdup(ru->ru_passwd);

			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, UID)) {
		if (new_user) {
			action = string("set uid to %u", ru->ru_uid);
		} else {
			action = string("change uid from %u to %u", ru->ru_pw->pw_uid, ru->ru_uid);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_pw->pw_uid = ru->ru_uid;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, GID)) {
		if (new_user) {
			action = string("set gid to %u", ru->ru_gid);
		} else {
			action = string("change gid from %u to %u", ru->ru_pw->pw_gid, ru->ru_gid);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_pw->pw_gid = ru->ru_gid;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, GECOS)) {
		if (new_user) {
			action = string("set GECOS to %s", ru->ru_gecos);
		} else {
			action = string("change GECOS to %s", ru->ru_gecos);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(ru->ru_pw->pw_gecos);
			ru->ru_pw->pw_gecos = strdup(ru->ru_gecos);

			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, DIR)) {
		if (new_user) {
			action = string("set home directory to %s", ru->ru_dir);
		} else {
			action = string("change home from %s to %s", ru->ru_pw->pw_dir, ru->ru_dir);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(ru->ru_pw->pw_dir);
			ru->ru_pw->pw_dir = strdup(ru->ru_dir);
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, MKHOME)) {
		action = string("create home directory");

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			/* mkdir $home; populate from $skel */
			if (mkdir(ru->ru_dir, 0700) == 0) {
				chown(ru->ru_dir, ru->ru_pw->pw_uid, ru->ru_pw->pw_gid);

				report_action(report, action, ACTION_SUCCEEDED);
			} else {
				report_action(report, action, ACTION_FAILED);
			}
		}

		if (ru->ru_skel) {
			action = string("populate home directory from %s", ru->ru_skel);

			if (dryrun) {
				report_action(report, string("populate home directory from %s", ru->ru_skel), ACTION_SKIPPED);
			} else {
				/* copy *all* files from ru_skel into ru_dir */
				if (_res_user_populate_home(ru->ru_dir, ru->ru_skel, ru->ru_pw->pw_uid, ru->ru_pw->pw_gid) == 0) {
					report_action(report, action, ACTION_SUCCEEDED);
				} else {
					report_action(report, action, ACTION_FAILED);
				}
			}
		}
	}

	if (res_user_different(ru, SHELL)) {
		if (new_user) {
			action = string("set login shell to %s", ru->ru_shell);
		} else {
			action = string("change shell from %s to %s", ru->ru_pw->pw_shell, ru->ru_shell);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(ru->ru_pw->pw_shell);
			ru->ru_pw->pw_shell = strdup(ru->ru_shell);
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, PWMIN)) {
		if (new_user) {
			action = string("set password minimum age to %u days", ru->ru_pwmin);
		} else {
			action = string("change password minimum age to %u days", ru->ru_pwmin);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_sp->sp_min = ru->ru_pwmin;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, PWMAX)) {
		if (new_user) {
			action = string("set password maximum age to %u days", ru->ru_pwmax);
		} else {
			action = string("change password maximum age to %u days", ru->ru_pwmax);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_sp->sp_max = ru->ru_pwmax;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, PWWARN)) {
		if (new_user) {
			action = string("set password expiry warning to %u days", ru->ru_pwwarn);
		} else {
			action = string("change password expiry warning to %u days", ru->ru_pwwarn);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_sp->sp_warn = ru->ru_pwwarn;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, INACT)) {
		if (ru->ru_inact) {
			action = string("deactivate account");
		} else {
			action = string("activate account");
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_sp->sp_inact = ru->ru_inact;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, EXPIRE)) {
		if (new_user) {
			action = string("set account expiration to %u", ru->ru_expire);
		} else {
			action = string("change account expiration to %u", ru->ru_expire);
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			ru->ru_sp->sp_expire = ru->ru_expire;
			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	if (res_user_different(ru, LOCK)) {
		size_t len = strlen(ru->ru_sp->sp_pwdp) + 2;
		char pbuf[len], *password;
		memcpy(pbuf + 1, ru->ru_sp->sp_pwdp, len - 1);

		if (ru->ru_lock) {
			action = string("lock account");
			pbuf[0] = '!';
			password = pbuf;
		} else {
			action = string("unlock account");
			password = pbuf + 1;
		}

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else {
			xfree(ru->ru_sp->sp_pwdp);
			ru->ru_sp->sp_pwdp = strdup(password);

			report_action(report, action, ACTION_SUCCEEDED);
		}
	}

	return report;
}

int res_user_is_pack(const char *packed)
{
	return strncmp(packed, RES_USER_PACK_PREFIX, RES_USER_PACK_OFFSET);
}

char* res_user_pack(const void *res)
{
	const struct res_user *ru = (const struct res_user*)(res);
	assert(ru);

	return pack(RES_USER_PACK_PREFIX, RES_USER_PACK_FORMAT,
	            ru->ru_enf,
	            ru->ru_name,   ru->ru_passwd, ru->ru_uid,    ru->ru_gid,
	            ru->ru_gecos,  ru->ru_shell,  ru->ru_dir,    ru->ru_mkhome,
	            ru->ru_skel,   ru->ru_lock,   ru->ru_pwmin,  ru->ru_pwmax,
	            ru->ru_pwwarn, ru->ru_inact,  ru->ru_expire);
}

void* res_user_unpack(const char *packed)
{
	struct res_user *ru = res_user_new(NULL);

	if (unpack(packed, RES_USER_PACK_PREFIX, RES_USER_PACK_FORMAT,
		&ru->ru_enf,
		&ru->ru_name,   &ru->ru_passwd, &ru->ru_uid,    &ru->ru_gid,
		&ru->ru_gecos,  &ru->ru_shell,  &ru->ru_dir,    &ru->ru_mkhome,
		&ru->ru_skel,   &ru->ru_lock,   &ru->ru_pwmin,  &ru->ru_pwmax,
		&ru->ru_pwwarn, &ru->ru_inact,  &ru->ru_expire) != 0) {

		res_user_free(ru);
		return NULL;
	}

	ru->key = string("res_user:%s", ru->ru_name);

	return ru;
}
